internal/docstore/docstore.go

```diff
@@ -268,10 +270,45 @@ func (l *ActionList) BeforeDo(f func(asFunc func(interface{}) bool) error) *Acti
 // attribute failures to specific actions; in such cases, the returned
 // ActionListError will have entries whose Index field is negative.
 func (l *ActionList) Do(ctx context.Context) error {
+	das, err := l.toDriverActions()
+	if err != nil {
+		return err
+	}
+	dopts := &driver.RunActionsOptions{
+		Unordered: l.unordered,
+		BeforeDo:  l.beforeDo,
+	}
+	alerr := ActionListError(l.coll.driver.RunActions(ctx, das, dopts))
+	if len(alerr) == 0 {
+		return nil // Explicitly return nil, because alerr is not of type error.
+	}
+	for i := range alerr {
+		alerr[i].Err = wrapError(l.coll.driver, alerr[i].Err)
+	}
+	return alerr
+}
+
+func (l *ActionList) toDriverActions() ([]*driver.Action, error) {
 	var das []*driver.Action
 	var alerr ActionListError
+	// Create a set of (document key, is Get action) pairs for detecting duplicates:
+	// an action list can have at most one get and at most one write for each key.
+	type keyAndKind struct {
+		key   interface{}
+		isGet bool
+	}
+	seen := map[keyAndKind]bool{}
```

Comment:
If the `key` is a pointer type, will this cause a false negative match? Should we make sure the key is not a pointer type?

Reply:
Done.